### PR TITLE
AZP: Test ucx_info from container w/o rdma devices.

### PR DESCRIPTION
**Pull request by petro-rudenko**
## What
Test ucx_info from container w/o rdma devices.

## Why ?
To justify #6210

### CODE REVIEW

File: buildlib/azure-pipelines-pr.yml [hunk 1]
```diff
@@ -163,6 +163,12 @@ stages:
               make -s -j `nproc`
             displayName: Build for $(CONTAINER)
 
+          - bash: |
+              set -eE
+              ../src/tools/info/ucx_info -e -u -t
```
Comment (evgeny-leksikov) [inline buildlib/azure-pipelines-pr.yml:168]: no need `-` for `t` value

**Comment by petro-rudenko** [comment]
@avildema @yosefe  - we need to add `librdacm` package to centos container to reproduce the issue

**Comment by avildema** [comment]
> @avildema @yosefe - we need to add `librdacm` package to centos container to reproduce the issue

Please use other container 

```    - container: rhel76
      image: rdmz-harbor.rdmz.labs.mlnx/swx-infra/x86_64/rhel7.6/builder:mofed-5.1-2.3.8.0
      options: -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools
```

or wait I will add it test and docker in other #6002

File: buildlib/azure-pipelines-pr.yml [hunk 2]
```diff
@@ -9,7 +9,7 @@ pr:
 resources:
   containers:
     - container: centos7
-      image: ucfconsort.azurecr.io/ucx/centos7:1
+      image: rdmz-harbor.rdmz.labs.mlnx/ucx/centos7-mofed5.x-cuda11.1:1
       endpoint: ucfconsort_registry
```
Comment (avildema) [inline buildlib/azure-pipelines-pr.yml:13]: please remove `endpoint`

**Comment by petro-rudenko** [comment]
```
2021-02-08T09:19:18.2753533Z [1612775958.271512] [6498d4fe7b84:17333:0]           ib_md.c:1700 UCX  ERROR ibv_open_device(mlx5_bond_0) failed: No such file or directory
2021-02-08T09:19:18.2833245Z [1612775958.279803] [6498d4fe7b84:17333:0]       rdmacm_cm.c:686  UCX  ERROR rdma_create_event_channel failed: No such file or directory
2021-02-08T09:19:18.2907693Z #
2021-02-08T09:19:18.2908922Z # UCP endpoint
2021-02-08T09:19:18.2909805Z #
2021-02-08T09:19:18.2910646Z #               peer: <no debug data>
2021-02-08T09:19:18.2912719Z #                 lane[0]:  2:self/memory.0 md[2]           -> md[2]/self     am am_bw#0
2021-02-08T09:19:18.2914514Z #                 lane[1]:  4:tcp/eth0.0 md[3]              -> md[3]/tcp      rma_bw#0
2021-02-08T09:19:18.2916358Z #                 lane[2]:  3:tcp/lo.0 md[3]                -> md[3]/tcp      rma_bw#1 wireup
2021-02-08T09:19:18.2917471Z #
2021-02-08T09:19:18.2918497Z #                tag_send: 0..<egr/short>..8185..<egr/bcopy>..8192..<rndv>..(inf)
2021-02-08T09:19:18.2919730Z #            tag_send_nbr: 0..<egr/short>..8185..<egr/bcopy>..262144..<rndv>..(inf)
2021-02-08T09:19:18.2920966Z #           tag_send_sync: 0..<egr/short>..8185..<egr/bcopy>..8192..<rndv>..(inf)
2021-02-08T09:19:18.2922062Z #
2021-02-08T09:19:18.2922937Z #                  rma_bw: mds [3] rndv_rkey_size 10
2021-02-08T09:19:18.2923829Z #
2021-02-08T09:19:18.3314649Z ##[section]Finishing: Test ucx_info
```

**Comment by petro-rudenko** [comment]
```
2021-02-08T09:49:36.6440457Z [ RUN      ] all/test_ucp_sockaddr_protocols_err.tag_rndv_unexp_put_scheme/4 <all/bidi_stop>
2021-02-08T09:49:37.0103918Z [     INFO ] < rdma_bind_addr(addr=10.224.36.142:0) failed: No such device >
2021-02-08T09:49:37.0105278Z [     INFO ] server listening on 10.224.36.142:46127
2021-02-08T09:49:37.0497650Z [swx-rdmz-ucx-new-02:28037:0:28037]      wireup.c:1053 Assertion `new_key->wireup_msg_lane == ((ucp_lane_index_t)-1)' failed
2021-02-08T09:49:38.7067093Z 
2021-02-08T09:49:38.7074186Z /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/ucp/wireup/wireup.c: [ ucp_wireup_check_config_intersect() ]
2021-02-08T09:49:38.7075208Z       ...
2021-02-08T09:49:38.7076314Z      1050     /* wireup lane hasn't been selected by the new configuration: only this
2021-02-08T09:49:38.7077231Z      1051      * function should select it */
2021-02-08T09:49:38.7078483Z      1052     ucs_assert(new_key->wireup_msg_lane == UCP_NULL_LANE);
2021-02-08T09:49:38.7079296Z ==>  1053 
2021-02-08T09:49:38.7079947Z      1054     /* set the correct WIREUP MSG lane in case of CM */
2021-02-08T09:49:38.7080929Z      1055     reuse_lane = reuse_lane_map[old_key->wireup_msg_lane];
2021-02-08T09:49:38.7081678Z      1056     if (reuse_lane != UCP_NULL_LANE) {
2021-02-08T09:49:38.7082226Z 
2021-02-08T09:49:38.8548201Z ==== backtrace (tid:  28037) ====
2021-02-08T09:49:38.8550476Z  0 0x000000000005d0e8 ucs_debug_print_backtrace()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/ucs/debug/debug.c:656
2021-02-08T09:49:38.8552822Z  1 0x00000000000a5814 ucp_wireup_check_config_intersect()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/ucp/wireup/wireup.c:1053
2021-02-08T09:49:38.8554393Z  2 0x00000000000a5814 ucp_wireup_init_lanes()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/ucp/wireup/wireup.c:1144
2021-02-08T09:49:38.8555632Z  3 0x00000000000a5d3e ucp_wireup_init_lanes_by_request()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/ucp/wireup/wireup.c:384
2021-02-08T09:49:38.8556880Z  4 0x00000000000a5d3e ucp_wireup_process_request()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/ucp/wireup/wireup.c:512
2021-02-08T09:49:38.8558232Z  5 0x00000000000a6c41 ucp_wireup_msg_handler()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/ucp/wireup/wireup.c:682
2021-02-08T09:49:38.8559462Z  6 0x0000000000022124 uct_iface_invoke_am()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/uct/base/uct_iface.h:663
2021-02-08T09:49:38.8560650Z  7 0x0000000000022124 uct_tcp_ep_comp_recv_am()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_ep.c:1199
2021-02-08T09:49:38.8561866Z  8 0x0000000000022124 uct_tcp_ep_progress_am_rx()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_ep.c:1333
2021-02-08T09:49:38.8563065Z  9 0x0000000000022124 uct_tcp_ep_progress_data_rx()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_ep.c:1430
2021-02-08T09:49:38.8564533Z 10 0x0000000000026b14 uct_tcp_iface_handle_events()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_iface.c:232
2021-02-08T09:49:38.8565724Z 11 0x000000000006af21 ucs_event_set_wait()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/ucs/sys/event_set.c:215
2021-02-08T09:49:38.8566919Z 12 0x0000000000026a24 uct_tcp_iface_progress()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_iface.c:249
2021-02-08T09:49:38.8568233Z 13 0x000000000003f05a ucs_callbackq_dispatch()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/ucs/datastruct/callbackq.h:211
2021-02-08T09:49:38.8569441Z 14 0x000000000003f05a uct_worker_progress()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/uct/api/uct.h:2436
2021-02-08T09:49:38.8570613Z 15 0x000000000003f05a ucp_worker_progress()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_worker.c:2428
2021-02-08T09:49:38.8571866Z 16 0x00000000008e3ebb ucp_test_base::entity::progress()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/ucp_test.cc:870
2021-02-08T09:49:38.8573124Z 17 0x00000000008e3f67 ucp_test::progress()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/ucp_test.cc:155
2021-02-08T09:49:38.8574587Z 18 0x00000000008e3fbb ucp_test::short_progress_loop()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/ucp_test.cc:163
2021-02-08T09:49:38.8575922Z 19 0x00000000008d0012 test_ucp_sockaddr_protocols::do_unexp_recv()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:1161
2021-02-08T09:49:38.8577436Z 20 0x00000000008d0012 ucs::entities_storage<ucp_test_base::entity>::receiver()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.h:202
2021-02-08T09:49:38.8579079Z 21 0x00000000008d0012 test_ucp_sockaddr_protocols::do_unexp_recv()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:1162
2021-02-08T09:49:38.8580491Z 22 0x00000000008d1012 test_ucp_sockaddr_protocols::test_tag_send_recv()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:1274
2021-02-08T09:49:38.8582010Z 23 0x00000000008d1012 test_ucp_sockaddr_protocols::test_tag_send_recv()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:1275
2021-02-08T09:49:38.8583569Z 24 0x00000000008b6200 test_ucp_sockaddr_protocols_err::test_tag_send_recv()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:1676
2021-02-08T09:49:38.8585087Z 25 0x00000000008b6200 test_ucp_sockaddr_protocols_err_tag_rndv_unexp_put_scheme_Test::test_body()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:1741
2021-02-08T09:49:38.8587173Z 26 0x00000000005de986 ucs::test_base::run()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.cc:365
2021-02-08T09:49:38.8588727Z 27 0x00000000005de986 ucs::test_base::TestBodyProxy()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.cc:391
2021-02-08T09:49:38.8590191Z 28 0x00000000005c0309 HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3562
2021-02-08T09:49:38.8591571Z 29 0x00000000005b454d testing::Test::Run()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3635
2021-02-08T09:49:38.8592916Z 30 0x00000000005b4625 testing::TestInfo::Run()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3812
2021-02-08T09:49:38.8594274Z 31 0x00000000005b478f testing::TestCase::Run()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3930
2021-02-08T09:49:38.8595603Z 32 0x00000000005b8f3d testing::internal::UnitTestImpl::RunAllTests()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:5808
2021-02-08T09:49:38.8596993Z 33 0x00000000005b9220 testing::internal::UnitTestImpl::RunAllTests()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:5725
2021-02-08T09:49:38.8598305Z 34 0x000000000054f038 RUN_ALL_TESTS()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest.h:20059
2021-02-08T09:49:38.8599457Z 35 0x000000000054f038 main()  /scrap/azure/agent-07/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/main.cc:102
2021-02-08T09:49:38.8600181Z 36 0x00000000000223d5 __libc_start_main()  ???:0
2021-02-08T09:49:38.8600813Z 37 0x000000000059e8bd _start()  ???:0
2021-02-08T09:49:38.8601336Z =================================
2021-02-08T09:49:38.8602130Z [swx-rdmz-ucx-new-02:28037:0:28037] Process frozen...
```

**Comment by petro-rudenko** [comment]
```
[2021-02-08T10:07:51.154Z] [  FAILED  ] 1 test, listed below:

[2021-02-08T10:07:51.154Z] [  FAILED  ] test_datatype.ptr_array_locked_perf, where TypeParam =  and GetParam() = 
```

**Comment by petro-rudenko** [comment]
bot:pipe:retest

File: buildlib/azure-pipelines-pr.yml [hunk 3]
```diff
@@ -163,6 +166,13 @@ stages:
               make -s -j `nproc`
             displayName: Build for $(CONTAINER)
 
+          - bash: |
+              set -eE
+              cd build
+              ./src/tools/info/ucx_info -e -u t
```
Comment (yosefe) [inline buildlib/azure-pipelines-pr.yml:172]: i'd suggest to save the output to file (|&tee info.txt) and grep for any warn/error

**Comment by petro-rudenko** [comment]
It's now fails, since there's errors:
```
[1612871744.804230] [87797e913a4d:17357:0]           ib_md.c:1700 UCX  ERROR ibv_open_device(mlx5_bond_0) failed: No such file or directory
[1612871744.812913] [87797e913a4d:17357:0]       rdmacm_cm.c:686  UCX  ERROR rdma_create_event_channel failed: No such file or directory
```
though ucx_info exits with 0

**Comment by yosefe** [comment]
@petro-rudenko what is the reason it fails?

**Comment by petro-rudenko** [comment]
> i'd suggest to save the output to file (|&tee info.txt) and grep for any warn/error

```
[1612871744.804230] [87797e913a4d:17357:0]           ib_md.c:1700 UCX  ERROR ibv_open_device(mlx5_bond_0) failed: No such file or directory
[1612871744.812913] [87797e913a4d:17357:0]       rdmacm_cm.c:686  UCX  ERROR rdma_create_event_channel failed: No such file or directory
```
Output of ucx_info. There's probably forwarded host network interface to a container (`mlx5_0_bond`)

**Comment by evgeny-leksikov** [comment]
> Output of ucx_info. There's probably forwarded host network interface to a container (`mlx5_0_bond`)

is this device in "down" state? 
currently we handle only `ENODEV`, should we add `ENOENT` too?

**Comment by petro-rudenko** [comment]
It's up on a host:
```
[peterr@sputnik2 ~]$ ibdev2netdev
mlx5_bond_0 port 1 ==> bond0 (Up)
```
not sure what's inside a docker

**Comment by petro-rudenko** [comment]
So: 
1. There's a host device in a container:
```
[root@3702748b0506 ~]# /hpc/local/bin/lshca
---------------------------------------------------------------------------------------------------
Dev #1
 Desc: Mellanox Technologies MT27710 Family [ConnectX-4 Lx]
 PN: =N/A=
 PSID: MT_2420110034
 SN: =N/A=
 FW: 14.28.2006
 Tempr: =N/A=
---------------------------------------------------------------------------------------------------
  PCI_addr   |    RDMA     | Net | Numa | LnkStat | IpStat | Link | Rate | LnkCapWidth  | HCA_Type
---------------------------------------------------------------------------------------------------
0000:02:00.0 | mlx5_bond_0 |     |  0   |  actv   |        | Eth  |  25  | =N/A= G=N/A= |  MT4117
0000:02:00.1 |             |     |  0   |         |        |      |      | =N/A= G=N/A= |
---------------------------------------------------------------------------------------------------
```
2. But `/dev/infiniband/uverbs0` and `/dev/infiniband/rdma_cm` are not forwarded. 
```
[root@3702748b0506 ~]# ibv_devinfo
Failed to open device
open("/dev/infiniband/uverbs0", O_RDWR|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```
3. There are next solutions:
3.1. Add to docker parameters `--device=/dev/infiniband/uverbs0 --device=/dev/infiniband/rdma_cm`
3.2. Ignore this device via UCX env parameters
3.3. Run this container not on mlx host (on AZP one - where there's no RDMA devices)

@yosefe @avildema WDYT ?

**Comment by yosefe** [comment]
can we handle the error in ucx by ignoring such device + ucs_diag() message?

**Comment by evgeny-leksikov** [comment]
> can we handle the error in ucx by ignoring such device + ucs_diag() message?

we can handler specific device error for MD but not for CM

**Comment by petro-rudenko** [comment]
Maybe change log level to warning and grep for just for error? That logs are a good indication that something wrong inside a container. It's very common to forget to add `--device=/dev/infiniband/uverbs0 --device=/dev/infiniband/rdma_cm` ucx may select non-RDMA transport and would be difficult to debug with release mode, what's going on.

**Comment by yosefe** [comment]
@petro-rudenko can you please convert this to a warning which will give the user a hint what is wrong?
and as you said, grep for ERROR only

File: src/uct/ib/rdmacm/rdmacm_cm.c [hunk 4]
```diff
@@ -677,7 +677,10 @@ UCS_CLASS_INIT_FUNC(uct_rdmacm_cm_t, uct_component_h component,
     if (self->ev_ch == NULL) {
         if (errno == ENODEV) {
             status  = UCS_ERR_NO_DEVICE;
-            log_lvl = UCS_LOG_LEVEL_DIAG;
+            log_lvl = UCS_LOG_LEVEL_WARN;
```
Comment (evgeny-leksikov) [inline src/uct/ib/rdmacm/rdmacm_cm.c:680]: this should be diag

**Review by yosefe — APPROVED** [review]

**Comment by petro-rudenko** [comment]
squashed

**Review by evgeny-leksikov — APPROVED** [review]

**Pull request merged**